src/query/api/v1/handler/placement/delete_test.go

```diff
@@ -33,119 +34,151 @@ import (
-				shard.NewShard(1).SetState(shard.Available),
-			})),
-		placement.NewInstance().SetID("host3").SetIsolationGroup("c").SetWeight(10).
-			SetShards(shard.NewShards([]shard.Shard{
-				shard.NewShard(1).SetState(shard.Available),
+			placement.NewInstance().SetID("host2").SetShards(shard.NewShards([]shard.Shard{
+				shard.NewShard(1).SetState(shard.Leaving),
 			})),
+		})
+
+		w = httptest.NewRecorder()
+		req = httptest.NewRequest(DeleteHTTPMethod, "/placement/host1", nil)
+		req = mux.SetURLVars(req, map[string]string{"id": "host1"})
+		require.NotNil(t, req)
+		mockPlacementService.EXPECT().Placement().Return(basePlacement, 0, nil)
+		handler.ServeHTTP(serviceName, w, req)
+
+		resp = w.Result()
+		body, err = ioutil.ReadAll(resp.Body)
+		require.NoError(t, err)
+		require.Equal(t, http.StatusBadRequest, resp.StatusCode)
+		require.Equal(t, `{"error":"instances [host2] do not have all shards available"}`+"\n", string(body))
+
+		// Test OK
+		basePlacement = basePlacement.SetReplicaFactor(2).SetMaxShardSetID(2).SetInstances([]placement.Instance{
+			placement.NewInstance().SetID("host1").SetIsolationGroup("a").SetWeight(10).
+				SetShards(shard.NewShards([]shard.Shard{
+					shard.NewShard(0).SetState(shard.Available),
+				})),
+			placement.NewInstance().SetID("host2").SetIsolationGroup("b").SetWeight(10).
+				SetShards(shard.NewShards([]shard.Shard{
+					shard.NewShard(0).SetState(shard.Available),
+					shard.NewShard(1).SetState(shard.Available),
+				})),
+			placement.NewInstance().SetID("host3").SetIsolationGroup("c").SetWeight(10).
+				SetShards(shard.NewShards([]shard.Shard{
+					shard.NewShard(1).SetState(shard.Available),
+				})),
+		})
+		if serviceName == M3AggServiceName {
+			// Need to be mirrored in M3Agg case
+			basePlacement.SetReplicaFactor(2).SetMaxShardSetID(2).SetInstances([]placement.Instance{
+				placement.NewInstance().SetID("host1").SetIsolationGroup("a").SetWeight(10).
+					SetShards(shard.NewShards([]shard.Shard{
+						shard.NewShard(0).SetState(shard.Available),
+						shard.NewShard(1).SetState(shard.Available),
+					})),
+				placement.NewInstance().SetID("host2").SetIsolationGroup("b").SetWeight(10).
+					SetShards(shard.NewShards([]shard.Shard{
+						shard.NewShard(0).SetState(shard.Available),
+						shard.NewShard(1).SetState(shard.Available),
+					})),
+			})
+		}
+
+		w = httptest.NewRecorder()
+		req = httptest.NewRequest(DeleteHTTPMethod, "/placement/host1", nil)
+		req = mux.SetURLVars(req, map[string]string{"id": "host1"})
+		require.NotNil(t, req)
+		mockPlacementService.EXPECT().Placement().Return(basePlacement, 1, nil)
+		mockPlacementService.EXPECT().CheckAndSet(gomock.Any(), 1).Return(nil)
+		handler.ServeHTTP(serviceName, w, req)
+
+		resp = w.Result()
+		body, err = ioutil.ReadAll(resp.Body)
+		require.NoError(t, err)
+		// require.Equal(t, http.StatusOK, resp.StatusCode)
+		// if serviceName == M3AggServiceName {
+
+		// } else {
```

Comment:
Still need the if/else commented out code?
Also: perhaps use backticks here for these literal requests?